### ISSUE TITLE
smartcard redirection not automatically activated if enforced by server

### ISSUE THREAD

**Issue by gbhauck** [FreeRDP/FreeRDP#10807]
**Description**
The attempt to connect to a Windows (10 or 11) PC fails due to an unknown device ID 0xffffffff if the computer is in a domain.

**To Reproduce**
Executing `xfreerdp /cert:ignore /d:HQ /u:XXXXXXXXXX /v:YYYYYYYYYYYYY`
results in (trace - log see attached file):
```
Password:        
[11:40:30:471] [8149:00001fd6] [ERROR][com.winpr.sspi.Kerberos] - [kerberos_AcquireCredentialsHandleA]: krb5glue_get_init_creds (Cannot find KDC for realm "HQ" [-1765328230])
[11:40:30:475] [8149:00001fd6] [ERROR][com.winpr.sspi.Kerberos] - [kerberos_AcquireCredentialsHandleA]: krb5glue_get_init_creds (Cannot find KDC for realm "HQ" [-1765328230])
[11:40:30:512] [8149:00001fd6] [WARN][com.freerdp.core.connection] - [rdp_client_connect_auto_detect]: expected messageChannelId=1008, got 1003
[11:40:30:512] [8149:00001fd6] [WARN][com.freerdp.core.license] - [license_read_binary_blob_data]: license binary blob::type BB_ERROR_BLOB, length=0, skipping.
[11:40:30:578] [8149:00001fd6] [WARN][com.freerdp.core.connection] - [rdp_client_connect_auto_detect]: expected messageChannelId=1008, got 1003
[11:40:30:648] [8149:00001fd6] [INFO][com.freerdp.gdi] - [gdi_init_ex]: Local framebuffer format  PIXEL_FORMAT_BGRX32
[11:40:30:649] [8149:00001fd6] [INFO][com.freerdp.gdi] - [gdi_init_ex]: Remote framebuffer format PIXEL_FORMAT_BGRA32
[11:40:30:670] [8149:00001fd6] [INFO][com.freerdp.channels.rdpsnd.client] - [rdpsnd_load_device_plugin]: [static] Loaded fake backend for rdpsnd
[11:40:30:670] [8149:00001fd6] [INFO][com.freerdp.channels.drdynvc.client] - [dvcman_load_addin]: Loading Dynamic Virtual Channel ainput
[11:40:30:670] [8149:00001fd6] [INFO][com.freerdp.channels.drdynvc.client] - [dvcman_load_addin]: Loading Dynamic Virtual Channel rdpgfx
[11:40:30:670] [8149:00001fd6] [INFO][com.freerdp.channels.drdynvc.client] - [dvcman_load_addin]: Loading Dynamic Virtual Channel rdpsnd
[11:40:30:936] [8149:00001fe4] [INFO][com.freerdp.channels.rdpsnd.client] - [rdpsnd_load_device_plugin]: [dynamic] Loaded fake backend for rdpsnd
[11:40:31:244] [8149:00001fe1] [WARN][com.freerdp.channels.rdpdr.client] - [devman_get_device_by_id]: could not find device ID 0xffffffff
[11:40:31:244] [8149:00001fe1] [ERROR][com.freerdp.channels.rdpdr.client] - [rdpdr_process_irp]: irp_new failed with 55!
[11:40:31:244] [8149:00001fe1] [ERROR][com.freerdp.channels.rdpdr.client] - [rdpdr_process_receive]: rdpdr_process_irp failed with error 55
[11:40:31:245] [8149:00001fe1] [ERROR][com.freerdp.channels.rdpdr.client] - [rdpdr_virtual_channel_client_thread]: rdpdr_process_receive failed with error 55!
[11:40:31:245] [8149:00001fd6] [ERROR][com.freerdp.core] - [checkChannelErrorEvent]: rdpdr_virtual_channel_client_thread reported an error. Error was 55
[11:40:31:246] [8149:00001fd6] [ERROR][com.freerdp.core] - [freerdp_check_event_handles]: checkChannelErrorEvent() failed - 0
[11:40:31:246] [8149:00001fd6] [INFO][com.freerdp.client.common] - [client_auto_reconnect_ex]: Network disconnect!
[11:40:31:246] [8149:00001fd6] [ERROR][com.freerdp.client.x11] - [xf_client_thread]: Failed to check FreeRDP file descriptor
[11:40:31:285] [8149:00001fd5] [ERROR][com.freerdp.core] - [freerdp_abort_connect_context]: ERRCONNECT_CONNECT_CANCELLED [0x0002000B]

```

**Expected behavior**
A freerdp session-window to be shown.

**Application details**
* FreeRDP version (`xfreerdp /version` → `This is FreeRDP version 3.9.0 (n/a)`); But I also had the same problem with all other 3.x versions
* Command line used: `xfreerdp /cert:ignore /d:HQ /u:XXXXXXXXXX /v:YYYYYYYYYYYYY`
* Output of `xfreerdp /buildconfig`:
```
This is FreeRDP version 3.9.0 (n/a)
Build configuration: BUILD_TESTING=OFF WINPR_HAVE_AIO_H=1 WINPR_HAVE_EXECINFO_BACKTRACE=1 WINPR_HAVE_EXECINFO_BACKTRACE_SYMBOLS=1 WINPR_HAVE_EXECINFO_BACKTRACE_SYMBOLS_FD=1 WINPR_HAVE_EXECINFO_HEADER=1 WINPR_HAVE_FCNTL_H=1 WINPR_HAVE_GETLOGIN_R=1 WINPR_HAVE_GETPWUID_R=1 WINPR_HAVE_INTTYPES_H=1 WINPR_HAVE_POLL_H=1 WINPR_HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIB=1 WINPR_HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIBS= WINPR_HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL=1 WINPR_HAVE_STDBOOL_H=1 WINPR_HAVE_STDINT_H=1 WINPR_HAVE_STRERROR_R=1 WINPR_HAVE_STRNDUP=1 WINPR_HAVE_SYSLOG_H=1 WINPR_HAVE_SYS_EVENTFD_H=1 WINPR_HAVE_SYS_FILIO_H= WINPR_HAVE_SYS_SELECT_H=1 WINPR_HAVE_SYS_SOCKIO_H= WINPR_HAVE_SYS_TIMERFD_H=1 WINPR_HAVE_TM_GMTOFF=1 WINPR_HAVE_UNISTD_H=1 WINPR_HAVE_UNWIND_H=1 WITH_AAD=ON WITH_ABSOLUTE_PLUGIN_LOAD_PATHS=ON WITH_ADD_PLUGIN_TO_RPATH=OFF WITH_ALSA=ON WITH_AVX2=ON WITH_AVX2_AVAILABLE=1 WITH_BINARY_VERSIONING=OFF WITH_CAIRO=ON WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CJSON_REQUIRED=OFF WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CLIENT_SDL=ON WITH_CLIENT_SDL2=ON WITH_CLIENT_SDL3=OFF WITH_CLIENT_SDL_AVAILABLE=1 WITH_CLIENT_SDL_VERSIONED=OFF WITH_CUPS=ON WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_CODECS=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_EVENTS=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SCHANNEL=OFF WITH_DEBUG_SDL_EVENTS=OFF WITH_DEBUG_SDL_KBD_EVENTS=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF_AVAILABLE=0 WITH_DEBUG_URBDRC=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DOCUMENTATION=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=OFF WITH_DSP_FFMPEG=OFF WITH_DSP_FFMPEG_AVAILABLE=0 WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FDK_AAC=ON WITH_FFMPEG=OFF WITH_FREERDP_DEPRECATED=OFF WITH_FREERDP_DEPRECATED_COMMANDLINE=OFF WITH_FULL_CONFIG_PATH=OFF WITH_FUSE=ON WITH_GFX_FRAME_DUMP=OFF WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=ON WITH_INTERNAL_MD4=OFF WITH_INTERNAL_MD5=OFF WITH_INTERNAL_RC4=OFF WITH_IPP=OFF WITH_JPEG=ON WITH_JSONC_REQUIRED=ON WITH_JSON_DISABLED=OFF WITH_KEYBOARD_LAYOUT_FROM_FILE=ON WITH_KRB5=ON WITH_KRB5_NO_NTLM_FALLBACK=OFF WITH_LAME=ON WITH_LIBRARY_VERSIONING=ON WITH_LIBRESSL=OFF WITH_LODEPNG=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_NATIVE_SSPI=OFF WITH_NEON=OFF WITH_OPENCL=OFF WITH_OPENH264=TRUE WITH_OPENH264=TRUE WITH_OPENH264_LOADING=OFF WITH_OPENSSL=ON WITH_OPUS=ON WITH_OSS=ON WITH_PCSC=ON WITH_PKCS11=ON WITH_PLATFORM_SERVER=ON WITH_POLL=ON WITH_PROFILER=OFF WITH_PROXY=ON WITH_PROXY_APP=ON WITH_PROXY_EMULATE_SMARTCARD=OFF WITH_PROXY_MODULES=ON WITH_PULSE=ON WITH_RDTK=ON WITH_SAMPLE=OFF WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SDL_IMAGE_DIALOGS=OFF WITH_SDL_LINK_SHARED=ON WITH_SERVER=ON WITH_SERVER_CHANNELS=ON WITH_SERVER_INTERFACE=ON WITH_SHADOW=ON WITH_SHADOW_MAC=ON WITH_SHADOW_X11=ON WITH_SMARTCARD_EMULATE=ON WITH_SMARTCARD_INSPECT=OFF WITH_SMARTCARD_PCSC=ON WITH_SOXR=ON WITH_SSE2=ON WITH_SWSCALE=OFF WITH_SYSTEMD=ON WITH_THIRD_PARTY=OFF WITH_TIMEZONE_COMPILED=OFF WITH_TIMEZONE_FROM_FILE=ON WITH_TIMEZONE_ICU=OFF WITH_UNICODE_BUILTIN=OFF WITH_URIPARSER=ON WITH_VAAPI=OFF WITH_VAAPI=OFF WITH_VAAPI_AVAILABLE=0 WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_VERBOSE_WINPR_ASSERT=ON WITH_VIDEO_FFMPEG=OFF WITH_VIDEO_FFMPEG=OFF WITH_VIDEO_FFMPEG_AVAILABLE=0 WITH_WAYLAND=ON WITH_WEBVIEW=OFF WITH_WINPR_DEPRECATED=OFF WITH_WINPR_JSON=ON WITH_WINPR_TOOLS=ON WITH_WIN_CONSOLE=ON WITH_X11=ON WITH_XCURSOR=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XTEST=ON WITH_XV=ON WITH_ZLIB=ON
Build type:          Release
CFLAGS:              -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wno-error=incompatible-pointer-types -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fPIC -Wall -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -g -fno-omit-frame-pointer
Compiler:            GNU, 14.2.1
Target architecture: x64
Keyboard Shortcuts:
	<Right CTRL>
		releases keyboard and mouse grab
	<CTRL>+<ALT>+<Return>
		toggles fullscreen state of the application
	<CTRL>+<ALT>+c
		toggles remote control in a remote assistance session
	<CTRL>+<ALT>+m
		minimizes the application
	Action Script
		Executes a predefined script on key press.
		Should the script not exist it is ignored.
		Scripts can be provided at the default location ~/.config/freerdp/action.sh or as command line argument /action:script:<path>
		The script will receive the current key combination as argument.
		The output of the script is parsed for 'key-local' which tells that the script used the key combination, otherwise the combination is forwarded to the remote.

```
* OS version connecting to (server side): Windows 10 Business 22H2
* log output from a run with `/log-level:trace 2>&1 | tee xfreerdp_trace.log`: 
[xfreerdp_trace.log](https://github.com/user-attachments/files/17587432/xfreerdp_trace.log)

**Environment:**
 - OS: Linux
 - Version/Distribution: Fedora Linux 41 (Workstation Edition); but same Problem since Version 39
 - Architecture: x86_64

**Note**
Same error reported in merge `[channels,settings] add a setting to ignore invalid devices #8911`: https://github.com/FreeRDP/FreeRDP/pull/8911#issuecomment-2441918384

The error also occurs if you use a *local* user account on the target machine.

**Comment by akallabeth** [comment]
Hmm, could you add `/smartcard` for a test?
This looks like the server ignores that the client does not announce device redirection

**Comment by akallabeth** [comment]
Also a full log with `/log-level:debug` could help

**Comment by gbhauck** [comment]
Thank you very much for looking into this!
Some points:

- Great, brilliant: adding `/smartcard` fixed the problem!
- I had added a trace log to the issue - is that enough or should I attach another log file with /smartcard option. If yes - debug or trace level ?
- For me, this /smartcard work-around is perfectly adequate 😊

**Comment by akallabeth** [comment]
@gbhauck no, I know what is going on now, so should be easy to fix.

**Comment by luxzg** [comment]
Hi everyone, I hope it's ok to comment on this issue as it seems to be a regression with the same issue.

I am on Debian Testing and latest Remmina from apt / repo (versions below), and I started having same "login + immediate logout" issue. Debug log showed same pattern as in this issue, and the same workaround actually fixed my connections.

This part of debug lead me here:
```
[11:40:31:244] [8149:00001fe1] [WARN][com.freerdp.channels.rdpdr.client] - [devman_get_device_by_id]: could not find device ID 0xffffffff
[11:40:31:244] [8149:00001fe1] [ERROR][com.freerdp.channels.rdpdr.client] - [rdpdr_process_irp]: irp_new failed with 55!
[11:40:31:244] [8149:00001fe1] [ERROR][com.freerdp.channels.rdpdr.client] - [rdpdr_process_receive]: rdpdr_process_irp failed with error 55
```

I got same `could not find device ID 0xffffffff` and `failed with 55`

Below are versions from my machine.

Remmina full version

```
remmina --full-version
remmina-Message: 09:53:42.551: Remmina does not log all output statements. Turn on more verbose output by using "G_MESSAGES_DEBUG=remmina" as an environment variable.
More info available on the Remmina wiki at:
https://gitlab.com/Remmina/Remmina/-/wikis/Usage/Remmina-debugging

org.remmina.Remmina - 1.4.40 (git n/a)

NAME                TYPE            DESCRIPTION                                                     PLUGIN AND LIBRARY VERSION
RDP                 Protocol        RDP - Remote Desktop Protocol                                   RDP plugin: 1.4.40 (Git n/a), Compiled with libfreerdp 3.17.2 (3.17.2), Running with libfreerdp 3.24.1 (rev 3.24.1), H.264 Yes
RDPF                File            RDP - RDP File Handler                                          RDP plugin: 1.4.40 (Git n/a), Compiled with libfreerdp 3.17.2 (3.17.2), Running with libfreerdp 3.24.1 (rev 3.24.1), H.264 Yes
RDPS                Preference      RDP - Preferences                                               RDP plugin: 1.4.40 (Git n/a), Compiled with libfreerdp 3.17.2 (3.17.2), Running with libfreerdp 3.24.1 (rev 3.24.1), H.264 Yes
VNC                 Protocol        Remmina VNC Plugin                                              1.4.40    
VNCI                Protocol        Remmina VNC listener Plugin                                     1.4.40    
glibsecret          Secret          Secured password storage in the GNOME keyring                   1.4.40    

Build configuration: HAVE_ARPA_INET_H=1 HAVE_ERRNO_H=1 HAVE_FCNTL_H=1 HAVE_NETDB_H=1 HAVE_NETINET_IN_H=1 HAVE_NETINET_TCP_H=1 HAVE_SYS_SOCKET_H=1 HAVE_SYS_UN_H=1 HAVE_TERMIOS_H=1 HAVE_UNISTD_H=1 WITH_AVAHI=ON WITH_CLEANUP=OFF WITH_FREERDP3=ON WITH_GCRYPT=ON WITH_ICON_CACHE=ON WITH_IPP=OFF WITH_KF5WALLET=ON WITH_KIOSK_SESSION=ON WITH_LIBRARY_VERSIONING=ON WITH_MANPAGES=ON WITH_NEWS=OFF WITH_PYTHONLIBS=ON WITH_SSE2=ON WITH_STATS=OFF WITH_TIP=OFF WITH_TRANSLATIONS=ON WITH_UPDATE_DESKTOP_DB=ON WITH_X2GO=ON
Build type:          None
CFLAGS:              -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/build/reproducible-path/remmina-1.4.40+dfsg=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -Wdate-time -D_FORTIFY_SOURCE=2 -Werror=incompatible-pointer-types -fPIC -Wall -g
Compiler:            GNU, 15.2.0
Target architecture: x64
```

From apt:
```
apt list --installed | grep freerd

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

libfreerdp-client3-3/testing,now 3.24.1+dfsg-1 amd64 [installed,automatic]
libfreerdp-server3-3/testing,now 3.24.1+dfsg-1 amd64 [installed,automatic]
libfreerdp3-3/testing,now 3.24.1+dfsg-1 amd64 [installed,automatic]

pt list --installed | grep remm

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

remmina-common/testing,now 1.4.40+dfsg-2 all [installed,automatic]
remmina-plugin-rdp/testing,now 1.4.40+dfsg-2 amd64 [installed,automatic]
remmina-plugin-secret/testing,now 1.4.40+dfsg-2 amd64 [installed,automatic]
remmina-plugin-vnc/testing,now 1.4.40+dfsg-2 amd64 [installed,automatic]
remmina/testing,now 1.4.40+dfsg-2 amd64 [installed]
```

Debian version:
```
lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux forky/sid
Release:        n/a
Codename:       forky
```

If I need to open a new issue or supply whole debug logs please let me know. But as I said, my logs are basically as the initial post shows, and when I enabled smart cards in the Remmina connection profiles they all started working. Before the change I would have more than 90% connections immediately end with "Signing out". One in many would actually work. Note that I don't use smart cards, but the Windows Server I'm connecting to does have OTP plugin setup, unsure how or if that's somehow affecting it.

I'll proceed with the workaround for now.

**Comment by akallabeth** [comment]
@luxzg can you reproduce with `xfreerdp` or `sdl-freerdp`?
these parts have not changed since the issue was closed, so maybe some setting on `remmina` end is wrong. (that test would help ruling that out)

**Comment by luxzg** [comment]
OK, sorry for not replying but I had issues testing this later on.
From the info I gathered recently it seems that this was the bug in Remimina:
https://gitlab.com/Remmina/Remmina/-/commit/ba0392c2

I have also found out that flatpak Remmina fixed this while Debian repo did not, package was later removed from Debian repos, but no new package was introduced. I'll keep using Flatpak alternatives (including newer Remmina, KRDC, GNOME Connections, etc).

Still, thanks for your reply!
